### PR TITLE
index: add trace2 region for clear skip worktree

### PR DESCRIPTION
**Pull request by Haizzz**
In large repository using sparse checkout, checking whether a file marked with skip worktree is present on disk and its skip worktree bit should be cleared can take a considerable amount of time. Add a trace2 region to surface this information.

cc: Timothy Jones timothy@canva.com
cc: Jeff Hostetler jeffhost@microsoft.com
cc: Jeff Hostetler <git@jeffhostetler.com>
cc: Derrick Stolee <derrickstolee@github.com>
cc: Taylor Blau <me@ttaylorr.com>

### CODE REVIEW

**Comment by dscho** [comment]
> cc Timothy Jones @zmthy

@Haizzz please note that you need to use the email address format, see e.g. https://github.com/git/git/pull/1362#issue-1408039312

File: sparse-index.c [hunk 1]
```diff
@@ -499,6 +499,7 @@ void clear_skip_worktree_from_present_files(struct index_state *istate)
 		return;
 
 restart:
+	trace2_region_enter("index", "clear_skip_worktree_from_present_files", istate->repo);
```
[the remark this answers is not in the document]
Comment (jeffhostetler) [reply sparse-index.c:502]: Yes, `trace2_region_enter()` and `trace2_region_leave()` calls must always be balanced.  They are essentially push and pop operations on a per-thread "region stack".  So please move the `trace2_region_enter()` above the `restart` label.

**Comment by jeffhostetler** [comment]
Let me offer a suggestion over your original code.  Yours will tell you the time
spent in that loop and that's fine and that may be all that you need.  Sometimes,
however, you want to know why one instance was so much slower than another.

In the code suggestion below, we keep track of the number of `path_found()`
calls -- these are essentially `lstat()` and are known to be expensive.  Let's
also keep track of whether we had to restart the loop.

I had to break the `if (ce_skip_worktree() && path_found())` into separate
parts, so the diff may look a little larger than it actually is.

```
	intmax_t count_path = 0;
	intmax_t count_restart = 0;
	...
	trace2_region_enter("index", "clear_skip_worktree_from_present_files", istate->repo);
restart:
	for (i = 0; i < istate->cache_nr; i++) {
		struct cache_entry *ce = istate->cache[i];

		if (ce_skip_worktree(ce)) {
			count_path++;
			if (path_found(ce->name, &last_dirname, &dir_len, &dir_found)) {
				if (S_ISSPARSEDIR(ce->ce_mode)) {
					ensure_full_index(istate);
					count_restart++;
					goto restart;
				}
				ce->ce_flags &= ~CE_SKIP_WORKTREE;
			}
		}
	}
	trace2_data_intmax("index", istate->repo,
			   "clear_skip_worktree_from_present_files/count_path",
			   count_path);
	trace2_data_intmax("index", istate->repo,
			   "clear_skip_worktree_from_present_files/count_restart",
			   count_restart);
	trace2_region_leave("index", "clear_skip_worktree_from_present_files", istate->repo);
	...
```

**Comment by Haizzz** [comment]
Thank you Jeff! I've taken on your suggestion with a small variable name change to match another case in the file

**Pull request closed without merging**